components/searchTodo.js

```diff
@@ -0,0 +1,28 @@
+import { searchTodoInput } from "./domElements.js";
+import { globalTodos, setTodos } from "../entry.js";
+import { renderTodoList } from "./domActions.js";
+
+function searchTodo() {
+  const searchedText = searchTodoInput.value.toLowerCase();
+
+  const searchedTodos = globalTodos.filter((todo) =>
+    todo.title.toLowerCase().includes(searchedText)
+  );
+
+  setTodos(searchedTodos);
+  renderTodoList();
+}
+
+export function debounceSearch(fn, delay = 500) {
+  let timer;
+
+  return function () {
+    clearTimeout(timer);
+
+    timer = setTimeout(() => {
+      fn();
+    }, delay);
+  };
+}
```

Comment:
Currently this function doesn't handle any incoming arguments. 

```suggestion
export function debounceSearch(fn, delay = 500) {
  let timer;

  return function (...args) {
    clearTimeout(timer);

    timer = setTimeout(() => {
      fn.apply(this, args);
    }, delay);
  };
}
```

Please note the first change, `return function (...args) {`
We are using the spread operator to collect all the arguments of the function into and array.

The second change, `fn.apply(this, args)`
This is truly a beautiful piece of code in a sense that, the usage of `apply` and `this` makes this function a generic function to be used in various of places. 
The `Function.prototype.apply` method, calls the provided `function` using a `this` and `argmuents` to that function. 

The value of `this` will be resolved in runtime and it will be an object based on the context that it was called on. 
For example, if we call the debounce method normally in window, `this` will be the window object. If we use the debounce method as a part of another object, `this` will be resolved to the object itself. 

We can also use `Function.prototype.call` instead of `apply`. The main difference is that arguments has to be passed as an Object mostly in the form of an array in apply(That's why we used spread operator to gather all the arguments of the function), and if we use `call` we have to pass it like normal arguments.

```js
fn.apply(this, ['a', 'b'])
fn.call(this, 'a', 'b')
```